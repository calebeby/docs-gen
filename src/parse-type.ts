import { Type, SymbolFlags, Symbol, ts, Node } from 'ts-simple-ast'

export enum NodeTypes {
  Literal = 0,
  Base = 1,
  Object = 4,
  Array = 5,
  Union = 6,
  Intersection = 7,
}

const parseKeyVal = (prop: Symbol, node: Node) => {
  const value = parseType(prop.getTypeAtLocation(node), node)
  const optional = prop.hasFlags(SymbolFlags.Optional)
  const parsed: {
    key: string
    optional: boolean
    value: ParsedType
    comment?: string
  } = {
    key: prop.getName(),
    optional,
    value:
      optional && value.type === NodeTypes.Union
        ? {
            ...value,
            // if it is optional exclude undefined from the type
            subTypes: value.subTypes.filter(
              sub =>
                !(sub.type === NodeTypes.Base && sub.value === 'undefined'),
            ),
          }
        : value,
  }
  const declaration = prop.getDeclarations()[0]
  if (declaration) {
    const commentRange = declaration.getLeadingCommentRanges()
    if (commentRange.length > 0) {
      parsed.comment = commentRange[0].getText().trim()
    }
  }
  return parsed
}

const parseObject = (t: Type, node: Node) => {
  const obj: ParsedObject = {
    type: NodeTypes.Object,
    keys: t.getProperties().map(p => parseKeyVal(p, node)),
  }
  const stringIndexType = t.getStringIndexType()
  const numberIndexType = t.getNumberIndexType()
  if (stringIndexType !== undefined) {
    obj.stringIndex = parseType(stringIndexType, node)
  }
  if (numberIndexType !== undefined) {
    obj.numberIndex = parseType(numberIndexType, node)
  }
  return obj
}

const parseArray = (t: Type, node: Node) =>
  parseType(t.getArrayType() as Type<ts.Type>, node)

interface ParsedTrue extends ParsedLiteral {
  value: 'true'
}

interface ParsedFalse extends ParsedLiteral {
  value: 'false'
}

const isTypeLiteral = (t: ParsedType): t is ParsedLiteral =>
  t.type === NodeTypes.Literal
const isTrueType = (t: ParsedType): t is ParsedTrue =>
  isTypeLiteral(t) && t.value === 'true'
const isFalseType = (t: ParsedType): t is ParsedFalse =>
  isTypeLiteral(t) && t.value === 'false'

const parseUnion = (t: Type, node: Node) => {
  const types = t.getUnionTypes().map(t => parseType(t, node))
  if (types.find(isTrueType) && types.find(isFalseType)) {
    return types
      .filter(type => !isTrueType(type) && !isFalseType(type))
      .concat({ type: NodeTypes.Base, value: 'boolean' })
  }
  return types
}

interface ParsedLiteral {
  type: NodeTypes.Literal
  value: string
}
interface ParsedBaseType {
  type: NodeTypes.Base
  value: 'string' | 'number' | 'boolean' | 'any' | 'null' | 'undefined'
}
export type ParsedType =
  | ParsedLiteral
  | ParsedBaseType
  | ParsedArray
  | ParsedUnion
  | ParsedObject
  | ParsedIntersection
interface ParsedArray {
  type: NodeTypes.Array
  subType: ParsedType
}
interface ParsedUnion {
  type: NodeTypes.Union
  subTypes: ParsedType[]
}
interface ParsedObjectKey {
  key: string
  optional: boolean
  comment?: string
  value: ParsedType
}
interface ParsedObject {
  stringIndex?: ParsedType
  numberIndex?: ParsedType
  type: NodeTypes.Object
  keys: ParsedObjectKey[]
}
interface ParsedIntersection {
  type: NodeTypes.Intersection
  subTypes: ParsedType[]
}

const isObject = (t: Type<ts.Type>): boolean => {
  if (t.isArray()) return false
  if (t.isObject()) return true
  const def = t.getDefault()
  return def ? isObject(def) : false
}

const parseIntersection = (t: Type, node: Node): ParsedType => {
  const subTypes = t.getIntersectionTypes()
  if (subTypes.every(isObject)) {
    const subObjects = subTypes.map(obj => parseObject(obj, node))
    const subKeys = subObjects.reduce<{ [key: string]: ParsedObjectKey }>(
      (acc, subObject) => {
        subObject.keys.forEach(k => (acc[k.key] = k))
        return acc
      },
      {},
    )
    return Object.assign({}, ...subObjects, {
      keys: Object.entries(subKeys).map(([, key]) => key),
    })
  }
  return {
    type: NodeTypes.Intersection,
    subTypes: subTypes.map(t => parseType(t, node)),
  }
}

export const parseType = (t: Type, node: Node): ParsedType => {
  if (node === undefined) {
    throw new Error('its bad' + t.getText())
  }
  // console.log(t.getText(), node.getText())
  if (t.isLiteral()) return { type: NodeTypes.Literal, value: t.getText() }
  if (t.isString()) return { type: NodeTypes.Base, value: 'string' }
  if (t.isNumber()) return { type: NodeTypes.Base, value: 'number' }
  if (t.isBoolean()) return { type: NodeTypes.Base, value: 'boolean' }
  if (t.isArray())
    return { type: NodeTypes.Array, subType: parseArray(t, node) }
  if (t.isUnion())
    return { type: NodeTypes.Union, subTypes: parseUnion(t, node) }
  if (t.isUndefined()) return { type: NodeTypes.Base, value: 'undefined' }
  if (t.isNull()) return { type: NodeTypes.Base, value: 'null' }
  if (t.getText() === 'any') return { type: NodeTypes.Base, value: 'any' }
  if (t.isIntersection()) return parseIntersection(t, node)
  if (isObject(t)) return parseObject(t, node)

  throw new Error(`Unrecognized type: ${t.getText()}`)
}
