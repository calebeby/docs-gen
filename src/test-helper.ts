import Project from 'ts-simple-ast'

export const parseSource = (text: string) => {
  const project = new Project()
  return project.createSourceFile('file.ts', text)
}
