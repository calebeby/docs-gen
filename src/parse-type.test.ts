import { parseSource } from './test-helper'
import { SyntaxKind } from 'ts-simple-ast'
import { parseType } from './parse-type'

const T = (s: string) =>
  parseSource(`type Foo = ${s}`)
    .getDescendantsOfKind(SyntaxKind.TypeAliasDeclaration)[0]
    .getType()

test('string literal', () => {
  expect(parseType(T("'hello'"))).toMatchSnapshot()
})
test('string', () => {
  expect(parseType(T('string'))).toMatchSnapshot()
})
test('number literal', () => {
  expect(parseType(T('4'))).toMatchSnapshot()
})
test('number', () => {
  expect(parseType(T('number'))).toMatchSnapshot()
})
test('any', () => {
  expect(parseType(T('any'))).toMatchSnapshot()
})
test('null', () => {
  expect(parseType(T('null'))).toMatchSnapshot()
})
test('undefined', () => {
  expect(parseType(T('undefined'))).toMatchSnapshot()
})
test('boolean literal', () => {
  expect(parseType(T('true'))).toMatchSnapshot()
})
test('boolean', () => {
  expect(parseType(T('boolean'))).toMatchSnapshot()
})

describe('object', () => {
  it('should parse correctly with comments', () => {
    expect(
      parseType(
        T(`{
        foo: number
        /* bar is
        a really
        long comment */
        bar: 10
        /* comment here */
        baz?: 'hello'
        // UTC
        date: string
      }`),
      ),
    ).toMatchSnapshot()
  })
  it('should parse string index type', () => {
    expect(
      parseType(
        T(`{
          [key: string]: number
          foo: number
          bar: string
        }`),
      ),
    ).toMatchSnapshot()
  })
  it('should parse number index type', () => {
    expect(
      parseType(
        T(`{
          [key: number]: number
          foo: number
          bar: string
        }`),
      ),
    ).toMatchSnapshot()
  })
})

test('array', () => {
  expect(parseType(T(`number[]`))).toMatchSnapshot()
})
describe('union', () => {
  it('should parse a basic union', () => {
    expect(parseType(T(`(number|'hello')[]`))).toMatchSnapshot()
  })
  it('should parse a union with a boolean', () => {
    expect(parseType(T(`(number|boolean)[]`))).toMatchSnapshot()
  })
})
describe('intersection', () => {
  test('object', () => {
    expect(
      parseType(
        T(`
{
  [key: number]: string
  foo: string
  bar: string
} &
{
  /* hello */
  bar?: number
  [key: number]: number
}`),
      ),
    ).toMatchSnapshot()
  })
  test('primitive', () => {
    expect(parseType(T(`number & 'hello'`))).toMatchSnapshot()
  })
})
