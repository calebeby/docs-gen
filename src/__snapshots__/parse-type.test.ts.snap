// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`any 1`] = `
Object {
  "type": 1,
  "value": "any",
}
`;

exports[`array 1`] = `
Object {
  "subType": Object {
    "type": 1,
    "value": "number",
  },
  "type": 5,
}
`;

exports[`boolean 1`] = `
Object {
  "type": 1,
  "value": "boolean",
}
`;

exports[`boolean literal 1`] = `
Object {
  "type": 0,
  "value": "true",
}
`;

exports[`intersection object 1`] = `
Object {
  "keys": Array [
    Object {
      "key": "foo",
      "optional": false,
      "value": Object {
        "type": 1,
        "value": "string",
      },
    },
    Object {
      "comment": "/* hello */",
      "key": "bar",
      "optional": true,
      "value": Object {
        "type": 1,
        "value": "number",
      },
    },
  ],
  "numberIndex": Object {
    "type": 1,
    "value": "number",
  },
  "type": 4,
}
`;

exports[`intersection primitive 1`] = `
Object {
  "subTypes": Array [
    Object {
      "type": 1,
      "value": "number",
    },
    Object {
      "type": 0,
      "value": "\\"hello\\"",
    },
  ],
  "type": 7,
}
`;

exports[`null 1`] = `
Object {
  "type": 1,
  "value": "null",
}
`;

exports[`number 1`] = `
Object {
  "type": 1,
  "value": "number",
}
`;

exports[`number literal 1`] = `
Object {
  "type": 0,
  "value": "4",
}
`;

exports[`object should parse correctly with comments 1`] = `
Object {
  "keys": Array [
    Object {
      "key": "foo",
      "optional": false,
      "value": Object {
        "type": 1,
        "value": "number",
      },
    },
    Object {
      "comment": "/* bar is
        a really
        long comment */",
      "key": "bar",
      "optional": false,
      "value": Object {
        "type": 0,
        "value": "10",
      },
    },
    Object {
      "comment": "/* comment here */",
      "key": "baz",
      "optional": true,
      "value": Object {
        "type": 0,
        "value": "\\"hello\\"",
      },
    },
    Object {
      "comment": "// UTC",
      "key": "date",
      "optional": false,
      "value": Object {
        "type": 1,
        "value": "string",
      },
    },
  ],
  "type": 4,
}
`;

exports[`object should parse number index type 1`] = `
Object {
  "keys": Array [
    Object {
      "key": "foo",
      "optional": false,
      "value": Object {
        "type": 1,
        "value": "number",
      },
    },
    Object {
      "key": "bar",
      "optional": false,
      "value": Object {
        "type": 1,
        "value": "string",
      },
    },
  ],
  "numberIndex": Object {
    "type": 1,
    "value": "number",
  },
  "type": 4,
}
`;

exports[`object should parse string index type 1`] = `
Object {
  "keys": Array [
    Object {
      "key": "foo",
      "optional": false,
      "value": Object {
        "type": 1,
        "value": "number",
      },
    },
    Object {
      "key": "bar",
      "optional": false,
      "value": Object {
        "type": 1,
        "value": "string",
      },
    },
  ],
  "stringIndex": Object {
    "type": 1,
    "value": "number",
  },
  "type": 4,
}
`;

exports[`string 1`] = `
Object {
  "type": 1,
  "value": "string",
}
`;

exports[`string literal 1`] = `
Object {
  "type": 0,
  "value": "\\"hello\\"",
}
`;

exports[`undefined 1`] = `
Object {
  "type": 1,
  "value": "undefined",
}
`;

exports[`union should parse a basic union 1`] = `
Object {
  "subType": Object {
    "subTypes": Array [
      Object {
        "type": 1,
        "value": "number",
      },
      Object {
        "type": 0,
        "value": "\\"hello\\"",
      },
    ],
    "type": 6,
  },
  "type": 5,
}
`;

exports[`union should parse a union with a boolean 1`] = `
Object {
  "subType": Object {
    "subTypes": Array [
      Object {
        "type": 1,
        "value": "number",
      },
      Object {
        "type": 1,
        "value": "boolean",
      },
    ],
    "type": 6,
  },
  "type": 5,
}
`;
